stolon/src/lib.rs

```diff
@@ -0,0 +1,89 @@
+use std::{net::SocketAddr, sync::Arc};
+
+use tokio::net::TcpStream;
+use tokio_socks::tcp::Socks5Stream;
+use tower::{Service, ServiceExt};
+
+pub use zebra_chain::{
+    parameters::Network, serialization::ZcashDeserialize, transaction::Transaction,
+};
+
+pub type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+// Tor's IsolateSOCKSAuth option (on by default) prevents sharing circuits
+// for streams for which different SOCKS authentication was provided.
+
+const SOCKS_USERNAME: &str = "stolon";
+fn socks_password() -> String {
+    use rand::distributions::{Alphanumeric, Distribution};
+    Alphanumeric
+        .sample_iter(rand::thread_rng())
+        .take(16)
+        .collect()
+}
+
+/// Send a transaction over Tor to the Zcash network, using DNS seeders to find peers.
+#[tracing::instrument(skip(transaction))]
+pub async fn send_transaction(
+    tor_addr: SocketAddr,
+    network: Network,
+    transaction: Arc<Transaction>,
+) -> Result<(), BoxError> {
+    // We want to ensure that we get a different peer from the DNS seeder for
+    // each call, so that fanout will send multiple times to different peers
+    // rather than multiple times to the same peer. However, Tor only sends a
+    // single IP address per RESOLVED cell, and caches DNS lookups. As a
+    // workaround, instead of looking up an IP and then connecting, try
+    // connecting directly to the DNS seeder, so that DNS resolution is
+    // performed by the exit. Since we use a new circuit for each call, we hit a
+    // different exit each time.
+    let seed_domain = match network {
+        Network::Mainnet => "mainnet.seeder.zfnd.org:8233",
+        Network::Testnet => "testnet.seeder.zfnd.org:18233",
+    };
+    let conn = Socks5Stream::connect_with_password(
+        tor_addr,
+        seed_domain,
+        SOCKS_USERNAME,
+        socks_password().as_str(),
+    )
+    .await?
```

Comment:
Should we wrap this request in a timeout?
What should we do if the connect times out or fails?

Reply:
If the connection fails, the error is propagated upwards to the caller.  There shouldn't be a timeout here or in any other part of the body of the function, but `send_transaction` and `send_transaction_to` should have a comment like

https://github.com/ZcashFoundation/zebra/pull/1014/files#diff-f2a0a2df0cb5a189c9df6ab45e6c2c34R29-R31

```
///
/// Note that this method does not implement any timeout behavior, so callers may
/// want to layer it with a timeout as appropriate for their application.
///
```
and there should be a timeout applied in `main.rs`.